ssz_serialization/merkleization.nim

```diff
@@ -887,7 +913,7 @@ func hashTreeRootAux[T](
         index = indexAt(i)
         indexLayer = log2trunc(index)
       if index == 1.GeneralizedIndex:
-        let contentsHash = chunkedHashTreeRoot(totalChunks, asSeq x)
+        let contentsHash {.noinit.} = chunkedHashTreeRoot(totalChunks, asSeq x)
```

Comment:
Why should `let foo {.noinit.} = bar` be any different than `let foo = bar`? Nim can know that it doesn't have to zero-init, because the init is happening right there.

Reply:
in general, Nim doesn't have a strict regime for these kinds of things - it just spams the code with zeroMem everywhere regardless if it's needed or not - this masks bugs and makes code work at the expense of performance.

for example, `let x = f()`, when used with RVO, will `zeroMem` x twice: once on the `let` side and once inside `f` so we need two `noinit`:s - often, it will also create 2 copies, ie the c code looks something like `MyType tmp; MyType x; memset(tmp, 0); f(&x); x = tmp;` to avoid aliasing issues and inside `f` we'll find `MyTtype tmp; memset(tmp, 0); ...; *Result = tmp`.